### ISSUE TITLE
 Generator of tiled wallpapers

### ISSUE THREAD

**Issue by peteroupc** [anufrievroman/waypaper#28]
Include a generator of images that have the feel of [classic tiled wallpapers](https://github.com/peteroupc/classic-wallpaper).  The following are suggestions on how the generator can generate these tiled wallpapers.

- Relief effect, based on a two-color (silhouette) image (such as one generated by thresholding an existing image).  This generates a three-color image.
- Dither to a limited color palette, such as a classic 16-color palette.
- 8x8 two-color tiled pattern.
- Convert to grayscale then shift the grayscale palette to a color gradient.
- Hue shifting.
- Tileable noise.
- Tileable procedural textures.
- Tileable arrangements of symbols or small images with partial transparency.
- Other procedural tileable patterns.
- Combinations of any of the foregoing.

None of the suggestions above should involve artificial intelligence.

**Comment by anufrievroman** [comment]
To be honest, the code for this kind of generator and its UI will probably be larger than the entire code for this application at the moment :D I mean, it seems a bit too complex for a very niche feature. Maybe you have some examples of programs that do that or even better some projects that could be used as a dependency?

**Comment by peteroupc** [comment]
My interest here is in simple programs that generate tiled wallpapers of abstract designs without AI, especially if they evoke the feel of classic wallpapers from the 90s and earlier.

There are several open-source wallpaper generators, many of which are in Python and include the following:

- https://github.com/search?q=generat+wallpaper+NOT+ai+language%3APython&type=repositories&l=Python
- https://github.com/adi1090x/canvas (just a shell script that uses ImageMagick)
- https://medium.com/swlh/wallpapers-with-python-621693567dba
- https://github.com/search?q=%22wallpaper+generator%22+NOT+ai&type=repositories&s=stars&o=desc

Many of these repositories could use the suggestions given in my previous message to generate new kinds of wallpapers.

A Python wallpaper generator can involve the use of ImageMagick (especially its `convert` tool) as well as Python's `pillow` library.  In addition, it's enough for this issue to support the following operations in its user interface:

- Relief effect, described above. (Input: An image.)
- Dither to the 16-bit Windows palette. (Input: An image.)
- Dither to the 226-bit Web safe palette. (Input: An image.)
- Hue shifting. (Input: An image; a hue from 0 to 360.)
- Convert to grayscale then shift the grayscale palette to a two-color gradient. (Input: An image; two colors.)

Other operations (described in my previous message) can be added later if they're easy enough to write, or Waypaper can add a button that generates a random wallpaper with the help of a third-party wallpaper generator.

**Comment by peteroupc** [comment]
Indeed, the last four of the five effects I listed (and several more) are possible in less than a hundred lines of Python code and the ImageMagick `convert` tool: https://github.com/peteroupc/peteroupc.github.io/blob/master/desktopwallpaper.py

**Comment by anufrievroman** [comment]
I checked a bit all the repositories, yes, it's very interesting.
Note to myself: [This is easy to add](https://github.com/timozattol/wallpaper-generator) to waypaper.

The conceptual problem I have is that if a make a generator, I won't be able to include a bunch of controls with it to substantially change the generated images. But, considering that each generator generate very similar images just color pallet (which is the main parameter) is different, I wonder if it's not easier to just pre-generate a bunch of images for all color pallets that I plan to supply as options... Anyways, I'll think about it.

**Comment by anufrievroman** [comment]
Finally, the way the project if going, I don't think this will be implemented. It should be implemented separately as backend and plugged into waypaper as such, i.e. the waypaper will stay as the frontend only. Thank you for the discussions. If somebody is interested in starting this tiling generator backend project, I'll be glad to advise on what need to be implemented.
